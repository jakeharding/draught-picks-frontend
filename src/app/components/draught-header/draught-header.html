<!-- Generated template for the DraughtHeaderComponent component -->
<ion-toolbar color="primary" mode="ios">
  <ion-buttons slot="start">
    <ion-back-button></ion-back-button>
  </ion-buttons>
  <ion-title><span class="pointer" (click)="home()">Draught Picks</span></ion-title>
  <ion-buttons slot="end">

    <ion-button ion-text color="light" tid="logout" float-right (click)="logout()" class="pointer" *ngIf="isLoggedIn">
      Logout
    </ion-button>
  </ion-buttons>
</ion-toolbar>
