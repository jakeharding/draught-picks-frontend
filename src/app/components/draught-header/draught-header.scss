draught-header.can-go-back ion-back-button.show-back-button {
  display: block;
}
