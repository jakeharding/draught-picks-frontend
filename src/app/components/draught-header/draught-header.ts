import { Component } from '@angular/core';
import { AuthProvider } from '../../services/auth/auth';
import { NavController } from '@ionic/angular';
import { ActivatedRoute } from '@angular/router';

/**
 * Generated class for the DraughtHeaderComponent component.
 *
 * See https://angular.io/api/core/Component for more info on Angular
 * Components.
 */
@Component({
  selector: 'draught-header',
  templateUrl: 'draught-header.html',
  styleUrls: ['./draught-header.scss']
})
export class DraughtHeaderComponent {

  isLoggedIn: boolean;

  constructor(public authProvider: AuthProvider, public navCtrl: NavController, public router: ActivatedRoute) {
    this.isLoggedIn = this.authProvider.isLoggedIn();
  }

  /**
   * logout function
   * No parameters
   * Sends current logged in user to sign in page
   */
  logout() {
    this.authProvider.clearToken();
    this.navCtrl.navigateRoot('sign-in');
  }

  /**
   * home function
   * No parameters
   * Sends current logged in user to the home page
   */
  async home() {
    if (this.authProvider.isLoggedIn()) {
      await this.navCtrl.navigateRoot('/tabs/home');
    } else {
      await this.navCtrl.navigateRoot('sign-in');
    }
  }


}
