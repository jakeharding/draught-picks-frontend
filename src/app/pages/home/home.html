<!--
  Generated template for the HomePage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <draught-header></draught-header>
</ion-header>

<ion-content class="ion-padding">
  <ion-row>
    <ion-col class="below-header" size="12" sizeSm="10" offsetSm="1" sizeMd="8" offsetMd="2" sizeLg="6" offsetLg="3" sizeXl="6" offsetXl="3">

      <ion-segment color="primary" [(ngModel)]="isRecentBeersSelected">
        <ion-segment-button tid="recentSegment" value="yes">
          Recent
        </ion-segment-button>
        <ion-segment-button tid="recommendedSegment" checked value="no">
          Recommended
        </ion-segment-button>
      </ion-segment>
      <label class="beerError" *ngIf="BACcalc">You have had {{totalCount}} beers in the past 3 hours. You might want to find a ride home.</label>
      <div [ngSwitch]="isRecentBeersSelected">

        <div *ngSwitchCase="'yes'">
          <div class="ion-text-center" *ngIf="recents?.length === 0">
            <p>You haven't had any beers yet! <a tid="toSearchPage" class="pointer" routerLink="/search">Search for what you are drinking.</a></p>
          </div>
          <div *ngIf="recents?.length > 0">

            <strong class="title">Your recent beers:</strong>
            <hr>
<!--            TODO use ionic 4 infinte scroll -->
            <div class="list-container" id="infinite-scroller1" appInfiniteScroller [immediateCallback]="false" [scrollCallback]="recentScrollCallback" [loadMore]="loadMoreRecent">
              <beer-list [beers]="recents"></beer-list>
            </div>
          </div>
        </div>
        <div *ngSwitchCase="'no'">
            <div class="ion-text-center" *ngIf="recommended?.length === 0">
              <p>Your recommended beers are coming soon! If you haven't created your beer profile yet,
                <a class="pointer" routerLink="/preferences">click here to fill it out now!</a></p>
            </div>
            <div *ngIf="recommended?.length > 0">
              <strong class="title">Your recommended beers:</strong>
              <hr>
              <div class="list-container" id="infinite-scroller2" appInfiniteScroller [immediateCallback]="false" [scrollCallback]="recScrollCallback" [loadMore]="loadMoreRecommended">
                <beer-list [beers]="recommended"></beer-list>
              </div>
            </div>
        </div>
      </div>

    </ion-col>
  </ion-row>
</ion-content>
