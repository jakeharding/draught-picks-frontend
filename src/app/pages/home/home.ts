import { Component } from '@angular/core';
import { NavController, NavParams } from '@ionic/angular';
import Beer from '../../models/Beer';
import { BeerProvider } from '../../services/beer/beer';
import { EMPTY } from 'rxjs';
import { BasePage } from '../BasePage';
import { LIMIT } from '../../app.component';


/**
 * Generated class for the HomePage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@Component({
  selector: 'page-home',
  templateUrl: 'home.html',
  styleUrls: ['./home.scss']
})
export class HomePage extends BasePage {
  recents: Array<Beer>;
  recommended: Array<Beer>;
  beerForABV: Array<Beer>;
  BACcalc: boolean;
  totalCount: number;
  isRecentBeersSelected = 'no';
  recommendedOffset: number;
  recentOffset: number;
  loadMoreRecommended: boolean;
  loadMoreRecent: boolean;
  recScrollCallback: () => void;
  recentScrollCallback: () => void;


  /**
   * HomePage constructor
   * Required constructor for the HomePage class, sets the offset, loadMoreRecommended, loadMoreRecent,
   * and recommended values
   */
  constructor(public navCtrl: NavController,
              public beerProvider: BeerProvider) {
    super('home');
    this.recommendedOffset = 0;
    this.recentOffset = 0;
    this.loadMoreRecommended = true;
    this.loadMoreRecent = true;
    this.recommended = [];
    this.recents = [];
  }

  /**
   * ionViewWillEnter function
   * No Parameters
   * Gets the recent and recommended beers and sets the LIMIT and offset values for the scrolling functionality.
   * Also calls the scrolling functionality
   */
  ionViewWillEnter() {
    const time = new Date();
    time.setHours(time.getHours() - 3);
    const pastTime =  new Date(time);

    this.beerProvider.recents({limit: LIMIT, offset: 0}).toPromise().then(results => {

      this.recents = results;
      this.beerForABV = this.recents.filter(beer => {
        beer.recents = beer.recents.filter( currentBeer => {

          return (new Date(currentBeer.created_at)) > pastTime;
        });
        return beer.recents.length > 0;
      });

      this.totalCount = 0;
      if ( this.beerForABV.length > 0) {
        this.totalCount = this.beerForABV.map((current) => current.recents.length).reduce((acc, c) => {
          return acc + c;
        });
      }
      this.BACcalc = this.totalCount >= 4;
    });

    this.beerProvider.recommended({limit: LIMIT, offset: 0}).toPromise().then(results => {
      this.recommended = results;
    });

  }

  /**
   * Get more recent beers if available.
   * @returns - recents
   */
  getRecentBeers(event) {
    if (this.loadMoreRecent) {
      const queryParams = {
        limit: LIMIT,
        offset: this.recentOffset
      };
      return this.beerProvider.recents(queryParams).toPromise().then(beers => this.processRecentBeers(beers, event));
    }
    return EMPTY;
  }

  /**
   * Get more recommended beers if available.
   * @returns - recommended
   */
  getRecommendedBeers(event) {
    if (this.loadMoreRecommended) {
      const queryParams = {
        limit: LIMIT,
        offset: this.recommendedOffset
      };
      return this.beerProvider.recommended(queryParams).toPromise().then(beers => this.processRecommendedBeers(beers, event));
    }
    return EMPTY;

  }

  /**
   * processRecommendedBeers function
   * Parameters: Beer list
   * increases the offset and loads more beers while scrolling through the beer list
   */
  private processRecommendedBeers = (beers, event) => {
    event.target.complete();
    if (beers.length === 0) {
      this.loadMoreRecommended = false;
      return;
    }
    this.recommendedOffset += LIMIT;
    this.recommended = this.recommended.concat(beers);
  }


  /**
   * processRecentBeers function
   * Parameters: Beer list
   * increases the offset and loads more beers while scrolling through the beer list
   */
  private processRecentBeers = (beers, event) => {
    event.target.complete();
    if (beers.length === 0) {
      this.loadMoreRecent = false;
      return;
    }
    this.recentOffset += LIMIT;
    this.recents = this.recents.concat(beers);
  }

}
