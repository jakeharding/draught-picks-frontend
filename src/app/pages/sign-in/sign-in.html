<!--
  Generated template for the SignInPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <draught-header></draught-header>
</ion-header>

<ion-content class="ion-padding ion-text-center">
  <ion-row>
    <ion-col size="12" sizeSm="10" offsetSm="1" sizeMd="8" offsetMd="2" sizeLg="6" offsetLg="3" sizeXl="4" offsetXl="4">
      <h1>Please Sign In</h1>
      <form [formGroup]="signInForm" (submit)="signIn()">
        <ion-item>
          <ion-label position="floating">Username or Email</ion-label>
          <ion-input tid="usernameField" formControlName="username" type="text"></ion-input>
        </ion-item>
        <br>
        <ion-item>
          <ion-label position="floating">Password</ion-label>
          <ion-input tid="passwordField" formControlName="password" type="password"></ion-input>
        </ion-item>
<!--        Work around for ion-button not submitting form when enter key is pressed-->
        <button type="submit" class="ion-page-hidden"></button>
        <ion-button
            tid="signInButton"
            (click)="signIn();"
            class="ion-margin-top"
            [disabled]="signInForm.invalid"
        >Sign In</ion-button>
        <div class="ion-text-center">
          <p>Do you not have an account?</p>
          <a routerLink="/register" routerDirection="forward">Click here to register</a>
        </div>
      </form>
    </ion-col>
  </ion-row>
</ion-content>
