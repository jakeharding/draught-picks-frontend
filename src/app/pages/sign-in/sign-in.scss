:host {
  ion-title{
    position: absolute;
    top: 0;
    left: 0;
    padding: 0 90px 1px;
    width: 100%;
    height: 100%;
    text-align: center;
    font-weight: bold;
  }
  ion-content{
    text-align: center;
  }

  a{
    cursor: pointer;
    text-decoration: none;
  }
  form{
    text-align: center;
  }
}
