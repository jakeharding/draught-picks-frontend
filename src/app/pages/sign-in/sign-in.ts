import { NavController } from '@ionic/angular';
import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { RegistrationPage } from '../registration/registration';
import { AuthProvider } from '../../services/auth/auth';
import { BasePage } from '../BasePage';
import { ToastProvider } from '../../services/toast/toast';

/**
 * Generated class for the SignInPage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@Component({
  selector: 'page-sign-in',
  templateUrl: 'sign-in.html',
  styleUrls: ['./sign-in.scss']
})
export class SignInPage extends BasePage {
  signInForm: FormGroup;
  goToRegistration: any;
  goToResetPassword;
  private username: string;
  private password: string;

  constructor(public navCtrl: NavController, public formBuilder: FormBuilder, public toastProvider: ToastProvider,
              public authProvider: AuthProvider) {
    super('sign-in');
    if (this.authProvider.isLoggedIn()) {
      this.navCtrl.navigateRoot('tabs');
    }
    // this.goToRegistration = RegistrationPage;
    this.signInForm = formBuilder.group({
      username: [this.username, Validators.compose([Validators.required])],
      password: [this.password, Validators.compose([Validators.required, Validators.minLength(4)])]
    });
  }

  /**
   * signIn function
   * No Parameters
   * Gets the username and password values and validates the information then sends
   * user to the home page
   */
  public signIn() {
    this.authProvider.signIn(this.signInForm.value).then(
      (response) => {
        this.authProvider.setToken(response.token);
        this.navCtrl.navigateRoot('/tabs/home');
      },
      () => {
        this.toastProvider.errorToast('Unable to sign you in. Have you verified your email address?');
      }
    );

  }
}

