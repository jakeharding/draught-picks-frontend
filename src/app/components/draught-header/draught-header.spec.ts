/**
 * draught-header.spec.ts
 *
 * Created by jake
 * Created on 11/15/18
 *
 * Test the DraughtHeaderComponent
 */
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { DraughtHeaderComponent } from './draught-header';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NavController } from '@ionic/angular';
import { AuthProvider } from '../../services/auth/auth';
import { SignInPage } from '../../pages/sign-in/sign-in';
import { HomePage } from '../../pages/home/home';

describe('DraughtHeaderComponent', () => {
  let component: DraughtHeaderComponent;
  let fixture: ComponentFixture<DraughtHeaderComponent>;

  const mockAuthProvider = {
    isLoggedIn: jest.fn(),
    clearToken: jest.fn(),
    provide: jest.fn()
  };

  const mockNavCtrl = {
    setRoot: jest.fn(),
    provide: jest.fn()
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [DraughtHeaderComponent],
      providers: [
        {provide: AuthProvider, useValue: mockAuthProvider},
        {provide: NavController, useValue: mockNavCtrl}
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(DraughtHeaderComponent);
    component = fixture.debugElement.componentInstance;
    jest.clearAllMocks();
  });

  it('should initialize the DraughtHeaderComponent', () => {
    expect(component).toBeDefined();
    expect(fixture).toMatchSnapshot();
  });

  test('logout() should call authProvider.clearToken and navCtrl.setRoot', () => {
    component.logout();
    expect(mockAuthProvider.clearToken).toHaveBeenCalledTimes(1);
    expect(mockNavCtrl.setRoot).toHaveBeenCalledWith(SignInPage);
  });

  test('home() should call authProvider.isLoggedIn and setRoot with HomePage when a user is logged in', () => {
    mockAuthProvider.isLoggedIn.mockReturnValue(true);
    component.home();
    expect(mockAuthProvider.isLoggedIn).toHaveBeenCalledTimes(1);
    expect(mockNavCtrl.setRoot).toHaveBeenCalledWith(HomePage);
  });

  test('home() should call authProvider.isLoggedIn and setRoot with SignInPage when a user is not logged in', () => {
    mockAuthProvider.isLoggedIn.mockReturnValue(false);
    component.home();
    expect(mockAuthProvider.isLoggedIn).toHaveBeenCalledTimes(1);
    expect(mockNavCtrl.setRoot).toHaveBeenCalledWith(SignInPage);
  });
});
