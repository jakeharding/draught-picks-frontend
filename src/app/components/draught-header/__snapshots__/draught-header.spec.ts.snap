// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraughtHeaderComponent should initialize the DraughtHeaderComponent 1`] = `
<draught-header
  authProvider={[Function Object]}
  isLoggedIn="undefined"
  navCtrl={[Function Object]}
>
  
  <ion-header>
    
    
    <ion-navbar
      color="primary"
    >
      
      
      <ion-row>
        
        
        <ion-col
          col-6=""
          offset-3=""
        >
          
          
          <ion-title
            class="pointer"
            text-center=""
          >
            Draught Picks
          </ion-title>
          
        
        </ion-col>
        
        
        <ion-col
          col-2=""
        >
          
          
          
          
        
        </ion-col>
        
      
      </ion-row>
      
    
    </ion-navbar>
    
  
  </ion-header>
  
</draught-header>
`;
